README.txt
; $Id: README.txt,v 1.1.4.2 2011/02/08 06:23:25 johnbarclay Exp $

================================================================================
This is a placeholder for the 7.x-2.x branch
================================================================================




CHANGELOG.txt
Additional test coverage and improvements of ldap_user test form including executing actions.
Restricted provision to ldap and provsion to drupal to one server configuration each
Made $ldap_user and $ldap_user_entry use more consistent
Missed some entries.  see git diffs.
Mix of work on drupal to ldap provisioning, fixing simpletests, and reworking mapping UI.
Issue #1543382 by larowlan.  Finished implementing synching of profile fields to drupal user.
Issue #1634930 by boran.  Added hook alter to allow intervention in ldap authentication->allowUser method.
Issue #1601222 by jbagley to fix call ldap provision function
Issue #1559088 by NimbyDagda.  LDAP SSO "noisily" checks the remote users headers for mod_auth_kerb.
Issue #1535994 by cgmonroe.  Fix for static caching of ldap_attributes_needed
Issue #1548102 by MJCO.  Mixed case Drupal role names failed.
Issue #1546968 by MJCO.  Fix for empty signature bug/problem.
Added Configuration and initial code for synch REST webservice for push ldap user create, synch, cancel, etc.
Merged 7.x-1.0-dev and 7.x-2.0-dev
Added ldap_user_puid, ldap_user_puid_property, ldap_user_puid_sid, and ldap_user_current_dn fields and attached to user object
Added ldap_user module with focus of user provisioning and updateing.  Will have functionality of ldap provision, ldap profile, ldap synch, etc.
Switch to 7.x-2.0 branch.
Issue #1412030 by srinivas.kakde.  corrected problem where authentication binding by user was preserverd in authorization, etc.
Issue #1441694 by haydeniv. fixed some needed include in ldap sso.
Issue #1450932 by hotspoons. untested kerberos for ldap sso.
Issue #1451778 by paulpaul.  bad variable used in dev verison.
Issue #1452116 by paulpaul.  fixed db fields too short for ldap authorization
Issue #1446768 by johnbarclay.  refined ldap server ->search() method error handling
Reworked LDAP Authorization OG to improve user interface and to support mapping to more OG roles than "member".
Reworked User Interface for LDAP Authorization.
Issue #1416738 by fearlsgroove.  disable editing on ldap profile fields synched from ldap
Issue #1328750#comment-5540222 by heydeniv.  misc ldap sso issues.
Issue #1328750#comment-5540318 by haydeniv. Fix for ldap authentication validation function return.
Issue #1430428 by resplin.  Added patch to with hook_ldap_server_in_use() so servers cannot be deleted or disabled when in use.
Issue #1421980 by clifmo.  Patch by chris.leversuch to fix ldap_feeds.info file
Issue #1341340 by marleythedod.  Views fix.
Fixed some error messaging in the test ldap server form.
Issue #1398204 by froeser.  Add try catch to check_plain on token array.
Issue #1396786 by thesaunterer.  patch to fix undefined token attributes.
Issue #1393196 by mattyohe. fixed stripping of quotes in return attributes.
Issue #1393188 by mattyohe.  lengthened size of ldap query filter field.
Issue #1357412 by nathangervais with patch.  ldap name transform php needed to be executed when binding with user ldap credentials.
Issue #1328750 by Gribnif, catch by memfis.  ldap_sso problems related to moving to own module.
Issue #1362480 by ybizeul to fix ldap profile warning/error.
Some quick fixes to menus, debug statements and ldap_sso.
Issue #1359222 by SangersDrupalDude.  theme_ldap_authorization_admin_index was sending empty array.
Issue #1359574 by heatherwoz.  spelling error.
Issue #1328750 by Gribnif. I forgot to remove hook_boot and hook_user_logout from ldap authentication.module
Issue #1359600 by heatherwoz.  Fixed code that set all fields with not null in schema to required.  Made exception for checkboxes.
Issue #1358174 by SangersDrupalDude.  Changed '0' to 0 to avoid schema errors.
Issue #1328750 by Gribnif.  Moved LDAP SSO to a separate module.  Details on issue queue item.
11/22/2011 Coder Review formatting and tweaks.
Issue #1328750. Not finished with issue, but added check for command line in hook_boot.  was breaking drush and other command line scripts.
Issue #1317816 by johnbarclay.  LDAP Authorization simpletest was broken.  Improved authorization form directions.
Issue #1329254 by Simon Georges: bad t() function call.
Issue #1284632 by iKevin and heatherwoz: fixed link to external url in logon help url.
Issue #1333714 by cmurph: LDAP Authentication: Add the ability to change user login description.
Added LDAP Profile to 7.x-1.x branch and did quite a bit of code cleanup.  Also added display of mapping
  example data on mapping page.  Limited servers being looked at for profile data to the servers user
  is authenticating to.
Issue #1334138 by benys: LDAP Authentication: Strip REMOTE_USER domain name
Issue #1322786 by heatherwoz: Feeds: Return only users of a certain role in LDAP Drupal User LDAP Entry Fetcher
Issue #1332410 by David Reid: improved usability of ldap_help and ldap server debug pages
Issue #1332350 by David Reid: mess from last commits removal of validation and switch to ldap type classes.
Issue #1308854 by trprinty patch by joericapens: fixed check of uid as numeric, not integer
Issue #1322180 by cmurph. logout error thrown.
Issue #1310318 by heatherwoz. array_merge() was making numeric ldap attributes fail for authorization
Issue #1316090 by geeve.  fixed ldap options call that was broke in previous change.
added classes for specific ldap implementations: http://drupal.org/node/1115704#comment-5103276

==================
7.x-1.0-beta5
==================
Updated documentation page http://drupal.org/node/997082
Fix token in ldap_authorization config page
Fixed ldap_help uninstall of straggling variable
Moved readme to http://drupal.org/node/1300810
Fixed bug with ldap_feeds users. Added documentation for one use case at http://drupal.org/node/1300812
Fixed a number of ldap_feeds issues. Both feed fetchers work and parser now. But not heavily tested and no simpletest coverage.
Update 7100 for ldap_query to add scope field to ldap_query table
Issue #1296994 by johnbarclay.  case sensitivity issue in ldap query.

ldap_query/LdapQuery.class.php
<?php
// $Id: LdapServer.class.php,v 1.5.2.1 2011/02/08 06:01:00 johnbarclay Exp $

/**
 * @file
 * Defines server classes and related functions.
 *
 */

/**
 * LDAP Server Class
 *
 *  This class is used to create, work with, and eventually destroy ldap_server
 * objects.
 *
 * @todo make bindpw protected
 */
class LdapQuery {
  // LDAP Settings

  public $query_numeric_id;
  public $qid;
  public $name;
  public $sid;
  public $status;

  public $baseDn = array();
  public $base_dn_str = NULL;
  public $filter;
  public $attributes_str = NULL;
  public $attributes = array();

  public $sizelimit = 0;
  public $timelimit = 0;
  public $deref = LDAP_DEREF_NEVER;
  public $scope = LDAP_SCOPE_SUBTREE;


  public $inDatabase = FALSE;
  public $detailedWatchdogLog = FALSE;


  /**
   * Constructor Method
   */
  function __construct($qid) {
    if (!is_scalar($qid)) {
      return;
    }

    $query_records = array();
    if (module_exists('ctools')) {
      ctools_include('export');
      $result = ctools_export_load_object('ldap_query', 'names', array($qid));
      if (isset($result[$qid])) {
        $query_record = $result[$qid];
        foreach ($query_record as $property_name => $value) {
          $this->{$property_name} = $value;
        }
      }
    }
    else {
      $select = db_select('ldap_query')
        ->fields('ldap_query')
        ->condition('ldap_query.qid',  $qid)
        ->execute();
      foreach ($select as $record) {
        $query_records[$record->qid] = $record;
      }
      if (!isset($query_records[$qid])) {
        $this->inDatabase = FALSE;
        return;
      }
      $query_record = $query_records[$qid];
      foreach ($this->fields() as $field_id => $field ) {
        if (isset($query_record->$field_id)) {
          $this->{$field['property_name']} = @$query_record->$field_id;
        }
      }
    }

    // special properties that don't map directly from storage and defaults
    $this->inDatabase = TRUE;
    $this->detailedWatchdogLog = variable_get('ldap_help_watchdog_detail', 0);

    $this->baseDn = $this->linesToArray($this->base_dn_str);
    $this->attributes = ($this->attributes_str) ? $this->csvToArray($this->attributes_str, TRUE) : array();

  }

  /**
   * Destructor Method
   */
  function __destruct() {

  }


  /**
   * Invoke Method
   */
  function __invoke() {

  }

//  function search($base_dn = NULL, $filter, $attributes = array(), $attrsonly = 0, $sizelimit = 0, $timelimit = 0, $deref = LDAP_DEREF_NEVER) {

  function query() {
    ldap_server_module_load_include('php', 'ldap_servers', 'LdapServer.class');
    $ldap_server = new LdapServer($this->sid);
    $ldap_server->connect();
    $ldap_server->bind();
    $results = array();

    $count = 0;
    // for debugging paginations, set: $ldap_server->searchPageStart = NULL, 0, 1, ... ; $ldap_server->searchPageEnd = NULL, 0, 1, ...;
    foreach ($this->baseDn as $base_dn) {
      $result = $ldap_server->search($base_dn, $this->filter, $this->attributes, 0, $this->sizelimit, $this->timelimit, $this->deref, $this->scope);
      if ($result !== FALSE && $result['count'] > 0) {
        $count = $count + $result['count'];
        $results = array_merge($results, $result);
      }
    }
    $results['count'] = $count;

    return $results;
  }

  /**
   * Error methods and properties.
   */

  protected $_errorMsg = NULL;
  protected $_hasError = FALSE;
  protected $_errorName = NULL;

  public function setError($_errorName, $_errorMsgText = NULL) {
    $this->_errorMsgText = $_errorMsgText;
    $this->_errorName = $_errorName;
    $this->_hasError = TRUE;
  }

  public function clearError() {
    $this->_hasError = FALSE;
    $this->_errorMsg = NULL;
    $this->_errorName = NULL;
  }

  public function hasError() {
    return ($this->_hasError || $this->ldapErrorNumber());
  }

  public function errorMsg($type = NULL) {
    if ($type == 'ldap' && $this->connection) {
      return ldap_err2str(ldap_errno($this->connection));
    }
    elseif ($type == NULL) {
      return $this->_errorMsg;
    }
    else {
      return NULL;
    }
  }

  public function errorName($type = NULL) {
    if ($type == 'ldap' && $this->connection) {
      return "LDAP Error: " . ldap_error($this->connection);
    }
    elseif ($type == NULL) {
      return $this->_errorName;
    }
    else {
      return NULL;
    }
  }

  public function ldapErrorNumber() {
   // if ($this->connection && ldap_errno($this->connection)) {
    //  return ldap_errno($this->connection);
   // }
   // else {
      return FALSE;
   // }
  }

  protected function linesToArray($lines) {
    $lines = trim($lines);
    if ($lines) {
      $array = preg_split('/[\n\r]+/', $lines);
      foreach ($array as $i => $value) {
        $array[$i] = trim($value);
      }
    }
    else {
      $array = array();
    }
    return $array;
  }

  protected function csvToArray($string, $strip_quotes = FALSE) {
    $items = explode(',', $string);
    foreach ($items as $i => $item) {
      $items[$i] = trim($item);
      if ($strip_quotes) {
        $items[$i] = trim($items[$i],'"');
      }
    }
    return $items;
  }

  public static function fields() {
    $fields = array(
      'query_numeric_id' => array(
          'property_name' => 'query_numeric_id',
          'schema' => array(
            'type' => 'serial',
            'unsigned' => TRUE,
            'not null' => TRUE,
            'description' => 'Primary ID field for the table.  Only used internally.',
            'no export' => TRUE,
          ),
        ),

      'qid' => array(
        'property_name' => 'qid',
        'schema' => array(
          'type' => 'varchar',
          'length' => 20,
          'description' => 'Machine name for query.',
          'not null' => TRUE,
          ),
        'form' => array(
          'field_group' => 'basic',
          '#type' => 'textfield',
          '#title' => t('Machine name for this query configuration.'),
          '#size' => 20,
          '#maxlength' => 20,
          '#description' => t('May only contain alphanumeric characters (a-z, A-Z, 0-9, and _)' ),
          '#required' => TRUE,
        ),
        'form_to_prop_functions' => array('trim'),
      ),

      'name' => array(
        'property_name' => 'name',
        'schema' => array(
          'type' => 'varchar',
          'length' => '60',
          'not null' => TRUE
        ),
        'form' => array(
          'field_group' => 'basic',
          '#type' => 'textfield',
          '#title' => t('Name'),
          '#description' => t('Choose a name for this query configuration.'),
          '#size' => 50,
          '#maxlength' => 255,
          '#required' => TRUE,
        ),
        'form_to_prop_functions' => array('trim'),
      ),

      'sid' => array(
        'property_name' => 'sid',
        'schema' => array(
          'type' => 'varchar',
          'length' => 20,
          'not null' => TRUE,
        ),
        'form' => array(
          'field_group' => 'basic',
          '#type' => 'radios',
          '#title' => t('LDAP Server used for query.'),
          '#required' => 1,
        ),
        'form_to_prop_functions' => array('trim'),
      ),

      'status' => array(
        'property_name' => 'status',
        'schema' => array(
          'type' => 'int',
          'size' => 'tiny',
          'not null' => TRUE,
          'default' => 0,
        ),
        'form' => array(
          'field_group' => 'basic',
          '#type' => 'checkbox',
          '#title' => t('Enabled'),
          '#description' => t('Disable in order to keep configuration without having it active.'),
        ),
        'form_to_prop_functions' => array('trim'),
      ),

      'base_dn_str' => array(
        'property_name' => 'base_dn_str',
        'schema' => array(
          'type' => 'text',
          'not null' => FALSE
        ),
        'form' => array(
          'field_group' => 'query',
          '#type' => 'textarea',
          '#title' => t('Base DNs to search in query.'),
          '#description' => t('Each Base DN will be queried and results merged. e.g. <code>ou=campus accounts,dc=ad,dc=uiuc,dc=edu</code>') . t('Enter one per line in case if you need more than one.'),
          '#cols' => 50,
          '#rows' => 6,
          '#required' => TRUE,
        ),
        'form_to_prop_functions' => array('trim'),
      ),

      'baseDn' => array(
        'property_name' => 'baseDn',
        'exportable' => FALSE,
      ),

      'filter' => array(
        'property_name' => 'filter',
        'schema' => array(
          'type' => 'text',
          'not null' => FALSE
        ),
        'form' => array(
          'field_group' => 'query',
          '#type' => 'textarea',
          '#title' => t('Filter'),
          '#description' => t('LDAP query filter such as <code>(objectClass=group)</code> or <code>(&(objectClass=user)(homePhone=*))
</code>'),
          '#cols' => 50,
          '#rows' => 1,
          '#required' => TRUE,
        ),
        'form_to_prop_functions' => array('trim'),
      ),

      'attributes_str' => array(
        'property_name' => 'attributes_str',
        'schema' => array(
          'type' => 'text',
          'not null' => FALSE
        ),
        'form' => array(
          'field_group' => 'query',
          '#type' => 'textarea',
          '#title' => t('Attributes to return.'),
          '#description' => t('Enter as comma separated list. DN is automatically returned. Leave empty to return all attributes. e.g. <code>objectclass,name,cn,samaccountname</code>'),
          '#cols' => 50,
          '#rows' => 6,
        ),
        'form_to_prop_functions' => array('trim'),
      ),

      'attributes' => array(
        'property_name' => 'attributes',
        'exportable' => FALSE,
      ),

      'sizelimit' => array(
        'property_name' => 'sizelimit',
        'schema' => array(
          'type' => 'int',
          'length' => 'small',
          'not null' => TRUE,
          'default' => 0,
        ),
        'form' => array(
          'field_group' => 'query_advanced',
          '#type' => 'textfield',
          '#title' => t('Size Limit of returned data'),
          '#description' => t('This limit may be already set by the ldap server.  0 signifies no limit'),
          '#size' => 7,
          '#maxlength' => 5,
          '#required' => TRUE,
        ),
        'form_to_prop_functions' => array('trim'),
      ),

      'timelimit' => array(
        'property_name' => 'timelimit',
        'schema' => array(
          'type' => 'int',
          'length' => 'small',
          'not null' => TRUE,
          'default' => 0,

        ),
        'form' => array(
          'field_group' => 'query_advanced',
          '#type' => 'textfield',
          '#title' => t('Time Limit in Seconds'),
          '#description' => t('The time limitset on this query.  This may be already set by the ldap server.  0 signifies no limit'),
          '#size' => 7,
          '#maxlength' => 5,
          '#required' => TRUE,
        ),
        'form_to_prop_functions' => array('trim'),
      ),

      'deref' => array(
        'property_name' => 'deref',
        'schema' => array(
          'type' => 'int',
          'size' => 'tiny',
          'not null' => TRUE,
          'default' => LDAP_DEREF_NEVER,
        ),
        'form' => array(
          'field_group' => 'query_advanced',
          '#type' => 'radios',
          '#title' => t('How aliases should be handled during the search.'),
          '#required' => 1,
          '#options' => array(
            LDAP_DEREF_NEVER => t('(default) aliases are never dereferenced.'),
            LDAP_DEREF_SEARCHING => t('aliases should be dereferenced during the search but not when locating the base object of the search.'),
            LDAP_DEREF_FINDING => t('aliases should be dereferenced when locating the base object but not during the search.'),
            LDAP_DEREF_ALWAYS => t('aliases should be dereferenced always.'),
          ),
        ),
        'form_to_prop_functions' => array('trim'),
      ),
     'scope' => array(
        'property_name' => 'scope',
        'schema' => array(
          'type' => 'int',
          'size' => 'tiny',
          'not null' => TRUE,
          'default' => LDAP_SCOPE_SUBTREE,
        ),
        'form' => array(
          'field_group' => 'query_advanced',
          '#type' => 'radios',
          '#title' => t('Scope of search.'),
          '#required' => 1,
          '#options' => array(
            LDAP_SCOPE_BASE => t('BASE. This value is used to indicate searching only the entry at the base DN, resulting in only that entry being returned (keeping in mind that it also has to meet the search filter criteria!).'),
            LDAP_SCOPE_ONELEVEL => t('ONELEVEL. This value is used to indicate searching all entries one level under the base DN - but not including the base DN and not including any entries under that one level under the base DN.'),
            LDAP_SCOPE_SUBTREE => t('SUBTREE. (default) This value is used to indicate searching of all entries at all levels under and including the specified base DN.'),
          ),
        ),
        'form_to_prop_functions' => array('trim'),
      ),

    );
    return $fields;
  }


}
